File: src/main/scala/io/phdata/streamliner/util/TemplateFunction.scala
/*
 * Copyright 2018 phData Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.phdata.streamliner.util

import io.phdata.streamliner.configuration.ColumnDefinition
import io.phdata.streamliner.configuration.Configuration
import io.phdata.streamliner.configuration.HadoopTable
import io.phdata.streamliner.configuration.Jdbc
import io.phdata.streamliner.configuration.SnowflakeTable
import io.phdata.streamliner.configuration.TableDefinition

object TemplateFunction {

  def mapDataType(
      column: ColumnDefinition,
      typeMapping: Map[String, Map[String, String]],
      storageFormat: String): String = {
    val cColumn = normalizeColumnDefinition(column)
    typeMapping.get(cColumn.dataType.toLowerCase) match {
      case Some(dataTypeMap) =>
        dataTypeMap.get(storageFormat.toLowerCase) match {
          case Some(dataType) => dataType
          case None =>
            throw new RuntimeException(
              s"No type mapping found for data type: '${column.dataType}' and storage format: $storageFormat in provided type mapping")
        }
      case None =>
        throw new RuntimeException(
          s"No type mapping found for data type: '${column.dataType}' in provided type mapping")
    }
  }

  def sqoopMapJavaColumn(tableDefinition: HadoopTable): Option[String] = {
    val map = tableDefinition.columns.flatMap { column =>
      val stringTypes =
        Seq("clob", "longvarbinary", "varbinary", "rowid", "blob", "nclob", "text", "binary")
      val intTypes = Seq("tinyint", "int", "smallint", "integer", "short")
      val floatTypes = Seq("float")
      val longDataTypes = Seq("bigint")
      val dataType = checkOracleNumberType(column).dataType.toLowerCase
      if (stringTypes.contains(dataType)) {
        Some(s"${column.destinationName}=String")
      } else if (floatTypes.contains(dataType)) {
        Some(s"${column.destinationName}=Float")
      } else if (intTypes.contains(dataType)) {
        Some(s"${column.destinationName}=Integer")
      } else if (longDataTypes.contains(dataType)) {
        Some(s"${column.destinationName}=Long")
      } else {
        None
      }
    }
    if (map.isEmpty) {
      None
    } else {
      Some(map.mkString(","))
    }
  }

  def cleanse(s: String): String = {
    val specialCharRegex = "(/|-|\\(|\\)|\\s|\\$)".r
    val specialChars = specialCharRegex.replaceAllIn(s.toLowerCase, "_")
    val dupsRegex = "(_{2,})".r
    val dups = dupsRegex.replaceAllIn(specialChars, "_")

    if (dups.startsWith("/") || dups.startsWith("_")) {
      dups.substring(1, dups.length)
    } else {
      dups
    }
  }

  def sourceColumns(configuration: Configuration, table: TableDefinition): String = {
    val jdbc = configuration.source.asInstanceOf[Jdbc]
    table.columns
      .map { column =>
        val f =
          if (jdbc.driverClass.get.contains("oracle") || jdbc.driverClass.get.contains("sqlserver")) {
            column.sourceName + " AS " + "\"" + column.destinationName + "\""
          } else {
            "`" + column.sourceName + "` AS " + column.destinationName
          }
        f
      }
      .mkString(",\n")
  }

  def columnOrCast(
      configuration: Configuration,
      table: TableDefinition,
      typeMapping: Map[String, Map[String, String]],
      sourceFormat: String,
      targetFormat: String): String =
    table.columns
      .map { column =>
        val sourceDataType = mapDataType(column, typeMapping, sourceFormat)
        val targetDataType = mapDataType(column, typeMapping, targetFormat)
        if (sourceDataType.equals(targetDataType)) {
          s"`${column.destinationName}`"
        } else {
          if (sourceDataType.equalsIgnoreCase("STRING") && targetDataType.equalsIgnoreCase(
              "DECIMAL")) {
            s"CAST(`${column.destinationName}` AS DECIMAL(${column.precision.get}, ${column.scale.get})) AS `${column.destinationName}`"
          } else {
            s"CAST(`${column.destinationName}` AS $targetDataType) AS `${column.destinationName}`"
          }
        }
      }
      .mkString(",\n")

  def primaryKeys(table: TableDefinition): String =
    table.primaryKeys.map(pk => s"`$pk`").mkString(",")

  /**
   * Normalize database-specific column definition properties.
   * @param columnDefinition ColumnDefinition to normalize
   * @return a new ColumnDefinition if changes are needed, the original ColumnDefinition otherwise
   */
  private[util] def normalizeColumnDefinition(
      columnDefinition: ColumnDefinition): ColumnDefinition = {
    val col = columnDefinition.copy(
      dataType = columnDefinition.dataType.toLowerCase.stripSuffix(" identity"))
    checkOracleNumberType(col)
  }

  private def checkOracleNumberType(columnDefinition: ColumnDefinition): ColumnDefinition = {
    val precision = columnDefinition.precision.getOrElse(0)
    val scale = columnDefinition.scale.getOrElse(0)
    if (columnDefinition.dataType.toUpperCase.equals("NUMBER")) {
      if (scale > 0) {
        columnDefinition.copy(dataType = "DECIMAL")
      } else if (precision > 19 && scale == 0) {
        columnDefinition.copy(dataType = "DECIMAL")
      } else if (precision >= 10 && precision <= 19 && scale == 0) {
        columnDefinition.copy(dataType = "BIGINT")
      } else if (precision == 0 && scale == -127) {
        columnDefinition.copy(dataType = "VARCHAR")
      } else {
        columnDefinition.copy(dataType = "INTEGER")
      }
    } else {
      columnDefinition
    }
  }

  def orderColumns(table: TableDefinition): Seq[ColumnDefinition] = {
    val primaryKeys = table.primaryKeys
    val primaryKeyColumnDefs =
      table.columns.filter(column => primaryKeys.contains(column.sourceName))
    val nonPrimaryKeyColumnDefs =
      table.columns.filter(column => !primaryKeys.contains(column.sourceName))
    primaryKeyColumnDefs ++ nonPrimaryKeyColumnDefs
  }
}

File: src/main/java/io/phdata/streamliner/schemadefiner/util/TemplateFunction.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
package io.phdata.streamliner.schemadefiner.util;

import io.phdata.streamliner.schemadefiner.model.ColumnDefinition;
import io.phdata.streamliner.schemadefiner.model.HadoopTable;
import io.phdata.streamliner.util.JavaHelper;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.apache.commons.lang3.StringUtils;

public class TemplateFunction {

  static final String[] zeroScaleSuffixes = {
    "_KEY",
    "_ID",
    "_NUM",
    "_CNUM",
    "_COUNT",
    "_COUNTER",
    "_CNT",
    "_DAY",
    "_DAY1",
    "_DAY2",
    "_WK",
    "_MTH",
    "_QTR",
    "_YR",
    "_PRECEDENCE",
    "_OFFSET",
    "_REL",
    "_SEQ",
    "_TS",
    "_AGO",
    "_ITEM",
    "_JLN"
  };

  public static String mapDataType(
      ColumnDefinition column,
      scala.collection.immutable.Map<String, scala.collection.immutable.Map<String, String>>
          typeMapping,
      String storageFormat) {
    Map<String, Map<String, String>> javaTypeMap = JavaHelper.convertScalaMapToJavaMap(typeMapping);
    ColumnDefinition cColumn = normalizeColumnDefinition(column, storageFormat);
    Map<String, String> dataTypeMap = javaTypeMap.get(cColumn.getDataType().toLowerCase());

    if (dataTypeMap == null || dataTypeMap.isEmpty()) {
      throw new RuntimeException(
          String.format(
              "No type mapping found for data type: '%s' in provided type mapping file",
              column.getDataType()));
    }
    String dataType = dataTypeMap.get(storageFormat.toLowerCase());
    if (dataType == null) {
      throw new RuntimeException(
          String.format(
              "No type mapping found for data type: '%s' and storage format: %s in provided type mapping",
              column.getDataType(), storageFormat));
    }
    return dataType;
  }

  private static ColumnDefinition checkOracleNumberType(
      ColumnDefinition columnDefinition, String format) {

    int precision = columnDefinition.getPrecision() == null ? 0 : columnDefinition.getPrecision();
    int scale = columnDefinition.getScale() == null ? 0 : columnDefinition.getScale();

    if (columnDefinition.getDataType().equalsIgnoreCase("NUMBER")) {
      if (scale > 0) {
        columnDefinition.setDataType("DECIMAL");
      } else if (precision > 19 && scale == 0) {
        columnDefinition.setDataType("DECIMAL");
      } else if (precision >= 10 && precision <= 19 && scale == 0) {
        columnDefinition.setDataType("BIGINT");
      } else if (precision == 0 && scale == -127 && !format.equalsIgnoreCase("SNOWFLAKE")) {
        columnDefinition.setDataType("VARCHAR");
      } else if (precision == 0 && scale == -127 && format.equalsIgnoreCase("SNOWFLAKE")) {
        columnDefinition.setDataType("NUMBER");
      } else {
        columnDefinition.setDataType("INTEGER");
      }
    } else {
      return columnDefinition;
    }
    return columnDefinition;
  }

  private static ColumnDefinition normalizeColumnDefinition(
      ColumnDefinition columnDefinition, String format) {
    columnDefinition.setDataType(stripSuffix(columnDefinition.getDataType()));
    return checkOracleNumberType(columnDefinition, format);
  }

  private static String stripSuffix(String dataType) {
    if (!dataType.toUpperCase().endsWith(" IDENTITY")) {
      return dataType.toUpperCase();
    } else {
      int index = dataType.toUpperCase().lastIndexOf(" IDENTITY");
      return dataType.substring(0, index);
    }
  }

  public static Boolean isZeroScaleSuffix(ColumnDefinition column) {
    return column.getDataType().equalsIgnoreCase("NUMBER")
        && column.getPrecision() == 0
        && (column.getScale() == -127 || column.getScale() == 0)
        && Arrays.stream(zeroScaleSuffixes)
            .anyMatch(suffix -> column.getSourceName().endsWith(suffix.toUpperCase()));
  }

  public static String sqoopMapJavaColumn(HadoopTable tableDefinition) {
    List<String> map =
        tableDefinition.getColumns().stream()
            .map(
                column -> {
                  String[] stringTypes = {
                    "clob", "longvarbinary", "varbinary", "rowid", "blob", "nclob", "text", "binary"
                  };
                  String[] intTypes = {"tinyint", "int", "smallint", "integer", "short"};
                  String[] floatTypes = {"float"};
                  String[] longDataTypes = {"bigint"};

                  List<String> stringList = Arrays.asList(stringTypes);
                  List<String> intList = Arrays.asList(intTypes);
                  List<String> floatList = Arrays.asList(floatTypes);
                  List<String> longList = Arrays.asList(longDataTypes);

                  String dataType =
                      checkOracleNumberType(column, "Hadoop").getDataType().toLowerCase();
                  if (stringList.contains(dataType)) {
                    return String.format("%s=String", column.getDestinationName());
                  } else if (floatList.contains(dataType)) {
                    return String.format("%s=Float", column.getDestinationName());
                  } else if (intList.contains(dataType)) {
                    return String.format("%s=Integer", column.getDestinationName());
                  } else if (longList.contains(dataType)) {
                    return String.format("%s=Long", column.getDestinationName());
                  } else {
                    return "";
                  }
                })
            .collect(Collectors.toList());
    if (map.isEmpty()) {
      return null;
    } else {
      return StringUtils.join(map, ",");
    }
  }
}
